tidy3d/components/data/monitor_data_n2f.py

```diff
@@ -36,24 +90,78 @@ def normalize(self, source_spectrum_fn: Callable[[float], complex]) -> AbstractN
-        eps_complex = medium.eps_model(frequency)
-        return medium.eps_complex_to_nk(eps_complex)
+        return medium.nk_model(frequency=self.f)
 
     @staticmethod
-    def k(frequency: float, medium: Medium) -> complex:
-        """Returns the complex wave number associated with the background medium."""
-        index_n, index_k = AbstractNear2FarData.nk(frequency, medium)
+    def propagation_factor(frequency: float, medium: Medium) -> complex:
+        """Complex valued wavenumber associated with a frequency and medium."""
+        index_n, index_k = medium.nk_model(frequency=frequency)
         return (2 * np.pi * frequency / C_0) * (index_n + 1j * index_k)
 
-    @staticmethod
-    def eta(frequency: float, medium: Medium) -> complex:
+    def k(self, medium: Medium) -> complex:
+        """Returns the complex wave number associated with the background medium."""
+        return self.propagation_factor(frequency=self.f, medium=medium)
+
+    def eta(self, medium: Medium) -> complex:
         """Returns the complex wave impedance associated with the background medium."""
-        eps_complex = medium.eps_model(frequency)
+        eps_complex = medium.eps_model(frequency=self.f)
         return ETA_0 / np.sqrt(eps_complex)
 
+    def near_field_rad_vecs(self, medium: Medium) -> Tuple[np.ndarray, np.ndarray]:
+        """Near field radiation vectors."""
+        eta = self.eta(medium=medium)
+        e_theta = -(self.Lphi.values + eta * self.Ntheta.values)
+        e_phi = self.Ltheta.values - eta * self.Nphi.values
+        return e_theta, e_phi
+
+    @staticmethod
+    def propagation_phase(dist: Union[float, None], wave_number: complex) -> complex:
+        """Phase associated with propagation of a distance with a given wavenumber."""
+        if dist is None:
+            return 1.0
+        return -1j * wave_number * np.exp(1j * wave_number * dist) / (4 * np.pi * dist)
+
+    def fields_sph(self, r: float = None, medium: Medium = Medium(permittivity=1)) -> xr.Dataset:
+        """Get fields in spherical coordinates relative to the monitor's local origin
+        for all angles and frequencies specified in :class:`Near2FarAngleMonitor`.
+        If the radial distance ``r`` is provided, a corresponding phase factor is applied
+        to the returned fields.
+
+        Parameters
+        ----------
+        r : float = None
+            (micron) radial distance relative to the monitor's local origin.
+        medium : :class:`.Medium`
+            Background medium in which to radiate near fields to far fields.
+            Default: free space.
+
+        Returns
+        -------
+        ``xarray.Dataset``
+            xarray dataset containing
+            (``E_r``, ``E_theta``, ``E_phi``, ``H_r``, ``H_theta``, ``H_phi``)
+            in polar coordinates.
+        """
+
+        # assemble E felds
+        e_theta_near, e_phi_near = self.near_field_rad_vecs(medium=medium)
```

Comment:
I think these are still far fields, just normalized by the phase factor. For example, if we pass `r=None`, the result is still a far field quantity. Maybe we could call these e.g. `e_theta_raw` or something?